### ISSUE TITLE
Accessibility of Snap for person with visual impairment

### ISSUE THREAD

**Issue by nathalierun** [jmoenig/Snap--Build-Your-Own-Blocks#1498]
My question is simple. We actually have in France (Education domain) a discussion about the accessibility of Scratch. http://revue.sesamath.net/spip.php?article908
Is it possible for a person with visual impairment to either create a program with Snap! or understand a program ?

**Comment by brianharvey** [comment]
We have made small steps in that direction, but I wouldn't say we're there yet.  The main thing is that we have a keyboard editor, allowing a script to be created entirely from the keyboard.  But we don't yet have audible feedback for graphical elements.

How to use:  First enable keyboard editing in the Settings (gear) menu.  Then...

     activate:
       - shift + click on a scripting pane's background
       - shift + click on any block
       - shift + enter in the IDE's edit mode

     stop editing:
       - left-click on scripting pane's background
       - esc

     navigate among scripts:
       - tab: next script
       - backtab (shift + tab): last script

     start editing a new script:
       - shift + enter

     navigate among commands within a script:
       - down arrow: next command
       - up arrow: last command

     navigate among all elements within a script:
       - right arrow: next element (block or input)
       - left arrow: last element

     move the currently edited script (stack of blocks):
       - shift + arrow keys (left, right, up, down)

     editing scripts:

       - backspace:
         * delete currently focused reporter
         * delete command above current insertion mark (blinking)
         * collapse currently focused variadic input by one element

       - enter:
         * edit currently focused input slot
         * expand currently focused variadic input by one element

       - space:
         * activate currently focused input slot's pull-down menu, if any
         * show a menu of reachable variables for the focused input or reporter

       - any other key:
         start searching for insertable matching blocks

       - in menus triggered by this feature:
         * navigate with up / down arrow keys
         * trigger selection with enter
         * cancel menu with esc

       - in the search bar triggered by this feature:
         * keep typing / deleting to narrow and update matches
         * navigate among shown matches with up / down arrow keys
         * insert selected match at the focus' position with enter
         * cancel searching and inserting with esc

    running the currently edited script:
        * shift + ctrl + enter simulates clicking the edited script with the mouse

**Comment by cycomachead** [comment]
What kind of visual impairment? 

Unfortunately, due to the implementation of Snap<em>!</em> there is zero compliace with WCAG 2.0 guidelines.

Total blindness is out of the question, at the moment. If you are interested in working on screen reader support I'd be happy to talk about the various techniques I've explored. :) Aside from descriptive output of the stage, it _should_ be possible to make Snap<em>!</em> work with screen readers, but it will be a significant amount of work. 

Otherwise, the best thing you can do is tweak the display settings of Snap<em>!</em>.

* The simplest thing is the zoom blocks setting which increases the size of the code blocks.
* You can also use your browser's zoom settings to make all the elements bigger.
* In the settings menu you might look at "Flat Design" mode - it's a much brighter UI, but the scripts area brackground is completely static and the contrast may be preferrable to some. (However, flat mode has it's own UI quirks too.)
* If you use Snap<em>!</em>'s "tables" display as a list output, then you can check "table lines" in 
the settings menu which great improve the readibility of tables for some.

There are some hidden options which may be useful:

If you shift-click the Settings (cog) menu, you get access to a lot of options which might be useful for someone with accessibility needs:
* Zebra-coloring: which will turn off the contrasting colors when nesting blocks.
* Blurred Shadows - this will make some shadow elements darker
* Drag Threshold. - perhaps more useful for motor disabilities, but this may be useful if you have low vision and can't distinguish how far you've dragged a set a block. (This controls what is considered a "drag" which is used for some stuff like creating new script variable instances.)

In Snap!, shift-click on the "Snap!" icon in the upper left and enter "Dev Mode"
Here you can tweak a few things which may be helpful to visually impaired users:
* Click on the grey background and select "touch screen settings" - this will make all the menus much bigger.
* You can change the color of many elements, but you could also inspect and edit the properities of any item. (If you do this often, it would be easier to just create a custom version of Snap<em>!</em> with separate settings, but this allows you to experiment.

**Comment by nathalierun** [comment]
Hello,

I'm sorry to thank you so late for those informations. I already use some of your advises. But they are going to be very usefull, specially during the courses with my students, because it was difficult for them to see some texts in the interface area.

**Comment by nathalierun** [comment]
cycomachead told me on his last post :
>  (If you do this often, it would be easier to just create a custom version of Snap! with separate settings, but this allows you to experiment.

How can I save my environment settings please ??
Because, every day,in classroom, I zoom on the blocks size, and I select "touch screen settings" in Dev Mode. I would like to save this in a file.
 Thank you by advance.

**Comment by jmoenig** [comment]
I'm sorry, but those are experimental and unsupported settings, and @cycomachead really shouldn't promote them or suggest creating custom versions of Snap for your class.

If you want every item in Snap to be larger, including blocks, buttons and menus, the "officially right" way to accomplish this is to do what you would do on any other website: Simply zoom the page using your browser's "zoom" option:

<img width="551" alt="zoom" src="https://cloud.githubusercontent.com/assets/2596285/20255496/65eb5dfe-aa3d-11e6-8e04-982dc29ad806.png">

You can either use the menu or the keyboard shortcut. Just remember to click into the browser's url bar if you want to use the keyboard shortcuts for zooming (so the browser gets focus instead of Snap).

Also, please make sure that visually impaired children or users adjust their screen resolution settings accordingly, so they see everything at at least normal size. Often times I see people scaling everything down to a tiny size simply to display more items on the screen.

**Comment by cycomachead** [comment]
I'm sorry, but not all visual impairments are the same! What works for one person doesn't work for everyone. Furthermore, equally zooming every element in a UI is _not_ the preferred option for many sighted users with low vision. Zooming as you suggest has other impairments on usability such as not being able to work with as much content. Iconography, colors, and locations all assist in helping people get to what they need. If all you need is a bigger font size, then the touch settings menus are a SIGNIFICANT improvement over a simple "Zoom" option provided by browsers. 



> Also, please make sure that visually impaired children or users adjust their screen resolution settings accordingly, so they see everything at at least normal size. Often times I see people scaling everything down to a tiny size simply to display more items on the screen.

No, no, no, no! It's fine to describe options to people, but we don't get to impose how people work or how people chose to manage their settings to aid any disabilities. 

> but those are experimental and unsupported settings, and @cycomachead really shouldn't promote them or suggest creating custom versions of Snap for your class.

As far as this goes -- perhaps I should have been more clear about them being unsupported, but I won't tell people they shouldn't modify their software to make it work for them! I _am_ a programmer in no small part because I learned to hack settings to make software easier to use. 

---

To actually solve the settings, there's a few things you could do but this simplest would probably be to have an extension for Chrome (or another browser) that automatically runs a script when you first load up Snap!
https://chrome.google.com/webstore/detail/tampermonkey/dhdgffkkebhmkfjojejmpbldmpobfkfo?hl=en

You'd need to test this, but all the script should need to call is:
```js
world.togglePreferences();
```

**Comment by jmoenig** [comment]
Wait! If you *only* want a bigger font size, please use your browser's settings to increase the *minimum font size*:

<img width="1096" alt="minfontsize" src="https://cloud.githubusercontent.com/assets/2596285/20256875/75f214d2-aa46-11e6-83e4-b518f95e206c.png">

This will increase the fonts in Snap, but leave everything else the same:

<img width="622" alt="minfontsizesnap" src="https://cloud.githubusercontent.com/assets/2596285/20256886/89a51592-aa46-11e6-9668-8808dfc9ef2f.png">

**Comment by cycomachead** [comment]
Ah, that works reasonably well, but for whatever reason Snap! doesn't respond to things that are in "zoom text only" mode. :/
